services/horizon/internal/resourceadapter/transaction.go

```diff
@@ -54,9 +54,51 @@ func PopulateTransaction(
 		}
 	}
 	dest.Signatures = row.Signatures
+
+	if row.HasPreconditions() {
+		dest.Preconditions = &protocol.TransactionPreconditions{}
+	}
+
 	if !row.TimeBounds.Null {
-		dest.ValidBefore = timeString(dest, row.TimeBounds.Upper)
-		dest.ValidAfter = timeString(dest, row.TimeBounds.Lower)
+		// Action needed in release: horizon-v3.0.0: remove ValidBefore and ValidAfter
+		dest.ValidBefore = timeString(row.TimeBounds.Upper)
+		dest.ValidAfter = timeString(row.TimeBounds.Lower)
+
+		if dest.Preconditions.Timebounds == nil {
+			dest.Preconditions.Timebounds = &protocol.TransactionPreconditionsTimebounds{}
+		}
+		dest.Preconditions.Timebounds.MaxTime = timeString(row.TimeBounds.Upper)
+		dest.Preconditions.Timebounds.MinTime = timeString(row.TimeBounds.Lower)
```

Comment:
wait, erm, weird question and might be a knowledge gap. why not always create the structure?

```suggestion
		dest.Preconditions.Timebounds = &protocol.TransactionPreconditionsTimebounds{
			MaxTime: timeString(row.TimeBounds.Upper),
			MinTime: timeString(row.TimeBounds.Lower),
		}
```